In [1]:
import numpy as np 
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
pd.set_option('display.max_columns', None) #Configuro Pandas para ver todas las columnas

In [3]:
df = pd.read_excel("../data (archivos del codigo principal)/dp2015-13_Dataset.xls")

In [4]:
df.head()

City   City_Eng City_Short   NAds  Price_Median   Price_Mean  \
0  Amsterdam  Amsterdam        AMS   9520   3419.972640  3426.322536   
1     Athina     Athens        ATH  10782   2064.004096  2109.106052   
2  Barcelona  Barcelona        BCN   5479   3140.000000  3267.621000   
3    Beograd   Belgrade        BEG  12797   1416.666667  1466.117265   
4     Berlin     Berlin        BER  16772   2150.338326  2313.968040   

   Area_Median  Area_Mean  Room_Median  Room_Mean  Euro_area  EU  Population  \
0         71.0  75.299475          3.0   2.838340          1   1      747093   
1         76.0  77.163884          3.0   2.870339          1   1      796442   
2         77.0  80.772769          3.0   2.794853          1   1     1615908   
3         58.0  60.236618          2.0   2.325467          0   0     1275000   
4         75.0  82.349263          2.5   2.579394          1   1     3431675   

   City_Area       Density        GDP_PC    GDP_PC_PPS    GDP_PC2008  \
0     165.90   4503.273056  49900.000000  49900.000000  55200.000000   
1     358.60   2220.976018  31900.000000  31900.000000  25400.000000   
2     100.10  16142.937063  30800.000000  30800.000000  28100.000000   
3     391.81   3254.128276   8453.264423  12037.008809   8453.264423   
4     892.00   3847.169283  24400.000000  24400.000000  26295.000000   

     GDP_PC2009    GDP_PC2010  Gini        HOR  Kearny_GCI2010   LRIR  \
0  52760.220994  53980.110497  25.5  21.000000            2.54   4.86   
1  25154.589372  24663.768116  32.9  51.900000            0.00   4.62   
2  27003.360098  27133.452691  33.0  68.100000            2.57   2.87   
3   7077.648127   7997.869718  28.2  97.424385            0.00  11.66   
4  26741.000000  27499.000000  30.0  12.800000            3.03   3.77   

   Inflation2010  Inflation2011      URate  MIR2009  MIR2010  Mortgage_PC2010  \
0       0.900000       2.500000   5.806452     5.26     4.57            38.01   
1       4.700000       3.100000  10.945455     3.41     3.79             7.12   
2       2.000000       3.100000  11.561947     2.52     2.54            14.79   
3      10.244648       7.004161  13.800000      NaN      NaN              NaN   
4       1.200000       2.500000  14.484000     4.29     3.70            14.09   

   Tppl1989_1993  Tppl1994_1998  Tppl1999_2002  Tppl2003_2006  Tppl2007_2009  \
0       702444.0       718119.0       734594.0       739104.0       747093.0   
1       772072.0       780572.0       789166.0       796442.0            NaN   
2      1643542.0      1508805.0      1503884.0      1578546.0      1615908.0   
3            NaN            NaN            NaN            NaN            NaN   
4      3465748.0      3458760.0      3388434.0      3387828.0      3431675.0   

   GDP_PC_PPS1989_1993  GDP_PC_PPS1994_1998  GDP_PC_PPS1999_2002  \
0                  NaN              28600.0              39400.0   
1                  NaN              13400.0              22000.0   
2                  NaN              17200.0              23400.0   
3                  NaN                  NaN                  NaN   
4                  NaN              19400.0              20900.0   

   GDP_PC_PPS2003_2006  GDP_PC_PPS2007_2009     CITIES  DemoDepend1989_1993  \
0              44800.0              49900.0  Amsterdam                 51.3   
1              27800.0              31900.0     Athina                 60.7   
2              25800.0              30800.0  Barcelona                 65.3   
3                  NaN                  NaN        NaN                  NaN   
4              21500.0              24400.0     Berlin                 51.9   

   DemoDepend1994_1998  DemoDepend1999_2002  DemoDepend2003_2006  \
0                 49.8                 48.4                 48.2   
1                 56.1                 52.0                 49.7   
2                 63.3                 61.7                 58.7   
3                  NaN                  NaN                  NaN   
4                 50.7        

In [5]:
'El dataset tiene {} filas y {} columnas.'.format(df.shape[0],df.shape[1])

'El dataset tiene 50 filas y 184 columnas.'

| **Columna**                      | **Significado**                                                                                                                                                                                                       | **Acción**                                  |
|-----------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------|
| **City** y **City_Eng**           | Nombre de la ciudad en idioma local e inglés.                                                                                                                                                                         | Eliminar City y cambiar el nombre de **City_Eng** a **City** |
| **City_Short**                    | Código de la ciudad.                                                                                                                                                                                                  | Eliminar.                                  |
| **NAds**                          | Número de anuncios publicados de apartamentos.                                                                                                                                                                       | Retener para análisis.                     |
| **Price_Median** y **Price_Mean** | Precio mediano y medio por m².                                                                                                                                                                                        | Usar para análisis económico y de precios (usar **Price Mean**). |
| **Area_Median** y **Area_Mean**   | Mediana y media del área total del apartamento.                                                                                                                                                                      | Retener para análisis de tamaño.           |
| **Room_Median** y **Room_Mean**   | Mediana y media del número de habitaciones.                                                                                                                                                                          | Retener para estadísticas de habitaciones. |
| **Euro_area** y **EU**            | **Euro Area**: países que usan el euro. **EU**: todos los estados miembros, ya sea que usen o no el euro.                                                                                                          | -                                           |
| **Population**, **City_Area** y **Density** | Población, área (km²) y densidad (personas/km²) de la ciudad.                                                                                                                                                       | Usar para análisis demográfico.            |
| **GDP_PC**, **GDP_PC_PPS**, **GDP_PC2008**, **GDP_PC2009**, **GDP_PC2010** | **GDP_PC**: Producto Interno Bruto per cápita. **GDP_PC_PPS**: Producto Interno Bruto per cápita ajustado por el poder de compra.                                                                                       | Elimino GDP_PC_PPS	GDP_PC2008	GDP_PC2009	GDP_PC2010	y me quedo con GDP_PC. Usar para análisis económico (usar **GDP_PC**) y de precios. |
| **Gini**                          | El coeficiente de Gini mide la desigualdad en el ingreso, con 0% reflejando igualdad perfecta y 100% máxima desigualdad.                                                                                           | -                                           |
| **HOR**                           | Homeownership rate (Tasa de propiedad de vivienda): porcentaje de hogares propietarios de la vivienda en la que viven.                                                                                              | -                                           |
| **Kearny_GCI2010**                | Kearney’s Global Cities Index: índice que clasifica ciudades por su capacidad para ser relevantes y competitivas a nivel global.                                                                                       | -                                           |
| **LRIR**, **MIR2009**, **MIR2010** | **LRIR**: Long-Run Interest Rate (Tasa de interés a largo plazo). **MIR**: Mortgage Interest Rate (Tasa de interés hipotecario) para los años 2009 y 2010.                                                             | -                                           |
| **Inflation2010** y **Inflation2011** | Tasas de inflación de 2010 y 2011.                                                                                                                                                                                    | -                                           |
| **URate**                         | Tasa de desempleo (de 0 a 100%).                                                                                                                                                                                       | Usar para análisis del mercado laboral.    |
| **Mortgage_PC2010**               | Hipotecas per cápita en 2010.                                                                                                                                                                                         | -                                           |
| **Latitude_deg, Latitude_min,Latitude_sec,Longitude_deg, Longitude_min, Longitude_sec, Lat, Lon**    | Columnas relacionadas con latitud y longitud,                                                                                                                                                                               | las elimino todas ya que no aportan al analisis que quiero realizar. |

**Valores nulos y duplicados**

In [8]:
lista_col_con_nulos = []
lista_col_20_nulos= []

for col in df:
    sum_nulls = df[col].isnull().sum() / len(df) * 100
    if sum_nulls != 0:
        lista_col_con_nulos.append(col)
    if sum_nulls > 20:
        lista_col_20_nulos.append(col) 

In [9]:
print(f'Number of columns in the dataframe                             = {df.shape[1]}')
print(f'Number of columns with nulls in the dataframe                  = {len(lista_col_con_nulos)}')
print(f'Number of columns with more than 20% of nulls in the dataframe = {len(lista_col_20_nulos)}')

Number of columns in the dataframe                             = 184
Number of columns with nulls in the dataframe                  = 150
Number of columns with more than 20% of nulls in the dataframe = 146


La mayoria de las columnas (150/184 tienen valores nulos, y de estas 150, 146 tienen más del 20% de valores nulos: son candidatas a ser eliminadas.

In [11]:
print("Las siguientes columnas contienen valores nulos y serán eliminadas:")
print(lista_col_20_nulos)

Las siguientes columnas contienen valores nulos y serán eliminadas:
['Tppl1989_1993', 'Tppl1994_1998', 'Tppl1999_2002', 'Tppl2003_2006', 'Tppl2007_2009', 'GDP_PC_PPS1989_1993', 'GDP_PC_PPS1994_1998', 'GDP_PC_PPS1999_2002', 'GDP_PC_PPS2003_2006', 'GDP_PC_PPS2007_2009', 'CITIES', 'DemoDepend1989_1993', 'DemoDepend1994_1998', 'DemoDepend1999_2002', 'DemoDepend2003_2006', 'DemoDepend2007_2009', 'DemoODepend1989_1993', 'DemoODepend1994_1998', 'DemoODepend1999_2002', 'DemoODepend2003_2006', 'DemoODepend2007_2009', 'Thh1989_1993', 'Thh1994_1998', 'Thh1999_2002', 'Thh2003_2006', 'Thh2007_2009', 'Ndwe1989_1993', 'Ndwe1994_1998', 'Ndwe1999_2002', 'Ndwe2003_2006', 'Ndwe2007_2009', 'Napart1989_1993', 'Napart1994_1998', 'Napart1999_2002', 'Napart2003_2006', 'Napart2007_2009', 'Nhouse1989_1993', 'Nhouse1994_1998', 'Nhouse1999_2002', 'Nhouse2003_2006', 'Nhouse2007_2009', 'Aphouse1989_1993', 'Aphouse1994_1998', 'Aphouse1999_2002', 'Aphouse2003_2006', 'Aphouse2007_2009', 'ApapartMincome1989_1993', 'Apa

In [12]:
df2 = df.drop(columns=lista_col_20_nulos,axis=1)

df2.head(5)


City   City_Eng City_Short   NAds  Price_Median   Price_Mean  \
0  Amsterdam  Amsterdam        AMS   9520   3419.972640  3426.322536   
1     Athina     Athens        ATH  10782   2064.004096  2109.106052   
2  Barcelona  Barcelona        BCN   5479   3140.000000  3267.621000   
3    Beograd   Belgrade        BEG  12797   1416.666667  1466.117265   
4     Berlin     Berlin        BER  16772   2150.338326  2313.968040   

   Area_Median  Area_Mean  Room_Median  Room_Mean  Euro_area  EU  Population  \
0         71.0  75.299475          3.0   2.838340          1   1      747093   
1         76.0  77.163884          3.0   2.870339          1   1      796442   
2         77.0  80.772769          3.0   2.794853          1   1     1615908   
3         58.0  60.236618          2.0   2.325467          0   0     1275000   
4         75.0  82.349263          2.5   2.579394          1   1     3431675   

   City_Area       Density        GDP_PC    GDP_PC_PPS    GDP_PC2008  \
0     165.90   4503.273056  49900.000000  49900.000000  55200.000000   
1     358.60   2220.976018  31900.000000  31900.000000  25400.000000   
2     100.10  16142.937063  30800.000000  30800.000000  28100.000000   
3     391.81   3254.128276   8453.264423  12037.008809   8453.264423   
4     892.00   3847.169283  24400.000000  24400.000000  26295.000000   

     GDP_PC2009    GDP_PC2010  Gini        HOR  Kearny_GCI2010   LRIR  \
0  52760.220994  53980.110497  25.5  21.000000            2.54   4.86   
1  25154.589372  24663.768116  32.9  51.900000            0.00   4.62   
2  27003.360098  27133.452691  33.0  68.100000            2.57   2.87   
3   7077.648127   7997.869718  28.2  97.424385            0.00  11.66   
4  26741.000000  27499.000000  30.0  12.800000            3.03   3.77   

   Inflation2010  Inflation2011      URate  MIR2009  MIR2010  Mortgage_PC2010  \
0       0.900000       2.500000   5.806452     5.26     4.57            38.01   
1       4.700000       3.100000  10.945455     3.41     3.79             7.12   
2       2.000000       3.100000  11.561947     2.52     2.54            14.79   
3      10.244648       7.004161  13.800000      NaN      NaN              NaN   
4       1.200000       2.500000  14.484000     4.29     3.70            14.09   

   Latitude_deg  Latitude_min  Latitude_sec  Longitude_deg  Longitude_min  \
0            52            23           0.0              4             54   
1            38             0           0.0             23             43   
2            41            23           0.0              2             11   
3            44            49           0.0             20             28   
4            52            31          20.0             13             17   

   Longitude_sec        Lat        Lon  
0            0.0  52.383333   4.900000  
1            0.0  38.000000  23.716667  
2            0.0  41.383333   2.183333  
3            0.0  44.816667  20.466667  
4           51.0  52.522222  13.297500

In [13]:
df2.shape

(50, 38)

Resulta un df con solo 34 columnas.
Ahora miro cuales son las columnas que quedaron y tienen menos del 20% de nulos 

In [15]:
lista_menos_20_nulos = []

lista_menos_20_nulos = list(set(lista_col_con_nulos) - set(lista_col_20_nulos))

lista_menos_20_nulos

['MIR2010', 'Inflation2011', 'Mortgage_PC2010', 'MIR2009']

In [16]:
#Calculo el numero de entradas nulas que tienen las columnas arriba detalladas
print(f'Nulls in MIR2009: {df2.MIR2009.isnull().sum()}')
print(f'Nulls in MIR2010: {df2.MIR2010.isnull().sum()}')
print(f'Nulls in Inflation2011: {df2.Inflation2011.isnull().sum()}')
print(f'Nulls in Mortgage_PC2010: {df2.Mortgage_PC2010.isnull().sum()}')

Nulls in MIR2009: 2
Nulls in MIR2010: 2
Nulls in Inflation2011: 6
Nulls in Mortgage_PC2010: 2


Chequeo si hay nulos

In [18]:
f'Numero de filas duplicadas en el dataset = {df2.duplicated().sum()}'

'Numero de filas duplicadas en el dataset = 0'

Para hacer el dataset mas facil de leer, elimino un par de columnas mas:

City y City_Short (el nombre de la ciudad en ingles es suficiente)
Hay 5 columnas relacionadas con el GDP, elimino GDP_PC_PPS	GDP_PC2008	GDP_PC2009	GDP_PC2010	y me quedo con GDP_PC
Elimino todas las columnas relacionadas con latitud y longitud.

In [20]:
df2.drop(['City',
            'City_Short',
            'Latitude_deg',
            'Latitude_min',
            'Latitude_sec',
            'Longitude_deg',
            'Longitude_min',
            'Longitude_sec',
            'Lat',
            'Lon',
            'GDP_PC_PPS',
            'GDP_PC2008',
            'GDP_PC2009',
            'GDP_PC2010',
            'NAds'],axis=1,inplace=True)

Finalmente, renombro las columnas al castellano

In [22]:
df2.columns

Index(['City_Eng', 'Price_Median', 'Price_Mean', 'Area_Median', 'Area_Mean',
       'Room_Median', 'Room_Mean', 'Euro_area', 'EU', 'Population',
       'City_Area', 'Density', 'GDP_PC', 'Gini', 'HOR', 'Kearny_GCI2010',
       'LRIR', 'Inflation2010', 'Inflation2011', 'URate', 'MIR2009', 'MIR2010',
       'Mortgage_PC2010'],
      dtype='object')

In [23]:
df2.rename(columns={'City_Eng':'Ciudad',
                      'Price_Median':'Mediana_Precios',
                      'Price_Mean':'Media_Precios',
                      'Area_Median':'Mediana_Area',
                      'Area_Mean':'Media_Area',
                      'Room_Median':'Mediana_habitaciones',
                      'Room_Mean':'Media_Habitaciones',
                      'Euro_area':'ZonaEuro',
                      'EU':'UnionEuropea',
                      'Population':'Poblacion',
                      'City_Area':'Area_Ciudad',
                      'Density':'Densidad Poblacional', 
                      'GDP_PC':'PIB_PC',
                      'HOR':'Tasa de Propiedad de Vivienda', 
                      'Kearny_GCI2010':'Indice Global de Ciudades de Kearny',
                      'LRIR':'Tasa de interés a largo plazo',
                      'Inflation2010':'Inflacion 2010',
                      'Inflation2011':'Inflacion 2011', 
                      'URate':'Tasa de Desempleo', 
                      'MIR2009':'Tasa de interés hipotecario 2009',
                      'MIR2010':'Tasa de interés hipotecario 2010',
       'Mortgage_PC2010':'Hipotecas per capita'},inplace=True)

df2.head()

Ciudad  Mediana_Precios  Media_Precios  Mediana_Area  Media_Area  \
0  Amsterdam      3419.972640    3426.322536          71.0   75.299475   
1     Athens      2064.004096    2109.106052          76.0   77.163884   
2  Barcelona      3140.000000    3267.621000          77.0   80.772769   
3   Belgrade      1416.666667    1466.117265          58.0   60.236618   
4     Berlin      2150.338326    2313.968040          75.0   82.349263   

   Mediana_habitaciones  Media_Habitaciones  ZonaEuro  UnionEuropea  \
0                   3.0            2.838340         1             1   
1                   3.0            2.870339         1             1   
2                   3.0            2.794853         1             1   
3                   2.0            2.325467         0             0   
4                   2.5            2.579394         1             1   

   Poblacion  Area_Ciudad  Densidad Poblacional        PIB_PC  Gini  \
0     747093       165.90           4503.273056  49900.000000  25.5   
1     796442       358.60           2220.976018  31900.000000  32.9   
2    1615908       100.10          16142.937063  30800.000000  33.0   
3    1275000       391.81           3254.128276   8453.264423  28.2   
4    3431675       892.00           3847.169283  24400.000000  30.0   

   Tasa de Propiedad de Vivienda  Indice Global de Ciudades de Kearny  \
0                      21.000000                                 2.54   
1                      51.900000                                 0.00   
2                      68.100000                                 2.57   
3                      97.424385                                 0.00   
4                      12.800000                                 3.03   

   Tasa de interés a largo plazo  Inflacion 2010  Inflacion 2011  \
0                           4.86        0.900000        2.500000   
1                           4.62        4.700000        3.100000   
2                           2.87        2.000000        3.100000   
3                          11.66       10.244648        7.004161   
4                           3.77        1.200000        2.500000   

   Tasa de Desempleo  Tasa de interés hipotecario 2009  \
0           5.806452                              5.26   
1          10.945455                              3.41   
2          11.561947                              2.52   
3          13.800000                               NaN   
4          14.484000                              4.29   

   Tasa de interés hipotecario 2010  Hipotecas per capita  
0                              4.57                 38.01  
1                              3.79                  7.12  
2                              2.54                 14.79  
3                               NaN                   NaN  
4                              3.70                 14.09

In [24]:
'El dataset tiene ahora {} filas y {} columnas.'.format(df2.shape[0],df2.shape[1])

'El dataset tiene ahora 50 filas y 23 columnas.'

Ahora observo el formato de los datos (fechas, strings, numericos, bool) para verificar que sean uniformes.

In [26]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Ciudad                               50 non-null     object 
 1   Mediana_Precios                      50 non-null     float64
 2   Media_Precios                        50 non-null     float64
 3   Mediana_Area                         50 non-null     float64
 4   Media_Area                           50 non-null     float64
 5   Mediana_habitaciones                 50 non-null     float64
 6   Media_Habitaciones                   50 non-null     float64
 7   ZonaEuro                             50 non-null     int64  
 8   UnionEuropea                         50 non-null     int64  
 9   Poblacion                            50 non-null     int64  
 10  Area_Ciudad                          50 non-null     float64
 11  Densidad Poblacional              

Las columnas Union Europea y ZonaEuro están como int64, pero podrían ser convertidas a bool si se consideran valores binarios, como por ejemplo 1 para verdadero (sí) y 0 para falso (no).


In [35]:
# Convertir las columnas 'Union Europea' y 'ZonaEuro' a tipo booleano
df2['UnionEuropea'] = df2['UnionEuropea'].astype(bool)
df2['ZonaEuro'] = df2['ZonaEuro'].astype(bool)


Si bien el tipo Object generalmente usado para cadenas de texto, también puede contener otros tipos de datos. El tipo String es específico para cadenas de texto y más eficiente para operaciones relacionadas con textos. Además, tiene algunos métodos adicionales optimizados para trabajar con strings.
Ya que la columna ciudad solo contiene valores texto, la convierto a tipo String.

In [38]:
df2['Ciudad'] = df2['Ciudad'].astype('string')